File: 2020/day06/day06.cpp
//
//  day06.cpp
//  day06
//
//  Created by Thomas CARTON on 06/12/2020.
//

#include <iostream>
#include <fstream>

#include <set>


int main(int argc, char *argv[])
{
    std::ifstream file("input.txt");

    // Part One:
    // count the number of questions to which anyone answered "yes"
    int anyoneCount = 0;

    std::set<char> unik;

    unsigned int k = 0;
    std::string line;
    while (std::getline(file, line))
    {
        if (k == 0)
        {

        }

        if (line.empty())
        {
            anyoneCount += unik.size();
            unik.clear();
        } else {
            unik.insert(line.begin(), line.end());
        }

        ++k;
    }
    anyoneCount += unik.size();

    // Part Two:
    // count the number of questions to which everyone answered "yes"
    int everyoneCount = 0;

    std::cout << "Part One: " << anyoneCount << '\n';

    std::cout << "Part Two: " << everyoneCount << '\n';

    return 0;
}

File: 2020/day06/day06.py
#!/usr/local/bin/python3.9
from collections import Counter

# Advent of Code 2020
# Day 6

def getInput(path: str = './input.txt') -> str:
    with open(path) as f:
        return f.read()

def main(input: str):
    anyone = 0
    everyone = 0

    for group in input.split('\n\n'):
        personCount = group.count('\n') + 1
        answers = group.replace('\n', '')

        anyone += len(set(answers))
        everyone += sum(count == personCount for _, count in Counter(answers).items())

    # Part One:
    # count the number of questions to which anyone answered "yes"
    print('Part One:', anyone)

    # Part Two:
    # count the number of questions to which everyone answered "yes"
    print('Part Two:', everyone)


if __name__ == '__main__':
    main(getInput())
